File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    boolean match(char a, char b)
    {
        if(b == '(' && a == ')')
            return true;
        else if(b == '[' && a == ']')
            return true;
        else if(b == '{' && a == '}')
            return true;
        return false;
    }
    public boolean isValid(String s) {
        ArrayDeque<Character> S = new ArrayDeque<>();
        for(int i=0; i<s.length(); i++)
        {
            char ch = s.charAt(i);
            if(ch == '(' || ch == '{' || ch == '[')
                S.push(ch);
            else{
                if(S.isEmpty() || !match(ch, S.peek()))
                    return false;
                else
                    S.pop();;
            }
        }
        if(S.isEmpty())
            return true;
        return false;
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        for i in s:
            if i in "({[": stack.append(i)
            else:
                if not stack: return False
                if (stack[-1] == '(' and i == ')') or \
                (stack[-1] == '[' and i == ']') or \
                (stack[-1] == '{' and i == '}'):
                    stack.pop()
                else: return False
        
        return not stack